tcp.go

```diff
@@ -33,6 +33,20 @@ func tryLinger(conn net.Conn, sec int) {
 	}
 }
 
+func tryKeepAlive(conn net.Conn, keepAlive bool) {
+	type canKeepAlive interface {
+		SetKeepAlive(bool) error
+	}
+
+	var _ canKeepAlive = &net.TCPConn{}
+
+	if keepALiveConn, ok := conn.(canKeepAlive); ok {
+		if err := keepALiveConn.SetKeepAlive(keepAlive); err != nil {
+			log.Warnf("Failed to enable TCP keepalive: %s", err)
+		}
+	}
```

Comment:
Maybe warn (error?) if we _can't_ set keepalives (bad interface)? That should never happen